[tool] Add --xcode-warnings-exceptions flag

loic-sharma opened the pull request:
Today, we treat Xcode warnings as errors on all iOS and macOS plugins.

However, the google_sign_in_ios plugin has dependencies with warnings. We're unable to suppress these warnings when using Swift Package Manager. As a workaround, we'll need to disable Xcode warnings for the google_sign_in_ios plugin. This will be done in a subsequent pull request (see https://github.com/flutter/packages/pull/7356).

This change introduces `--xcode-warnings-exceptions` to the `native-test` command. When running Xcode test, `GCC_TREAT_WARNINGS_AS_ERRORS` will be provided only if the plugin isn't on the exception list.

Additionally, plugins that are the exception list are excluded from the `xcode-analyze` command.

Part of https://github.com/flutter/flutter/issues/146904

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

vashworth reviewed (approved):
LGTM

stuartmorgan commented:
Do we get any meaningful benefit from running analysis without `GCC_TREAT_WARNINGS_AS_ERRORS` vs just not running analysis on excluded packages (using the existing `--exclude` flag)?

vashworth commented:
> Do we get any meaningful benefit from running analysis without `GCC_TREAT_WARNINGS_AS_ERRORS` vs just not running analysis on excluded packages (using the existing `--exclude` flag)?

That's a good point. It's doubtful that anyone will actually go read the warnings if they're not escalated as an error.

stuartmorgan commented:
I can pretty much guarantee that nobody will go look at passing output (except maybe retroactively if we later find that something has gone wrong, but that could also be checked locally pretty easily), what I was wondering was whether there's any scenario where analysis without -werror would fail but building wouldn't; i.e., some situation where this could still fail and alert us to a problem.

I don't know if there's such a thing as an analysis *error* that's not also a build error. My assumption is probably not, but wasn't sure.

loic-sharma commented on .ci/targets/macos_platform_tests.yaml:
```diff
@@ -11,11 +11,11 @@ tasks:
+    args: ["xcode-analyze", "--macos", "--macos-min-version=14.0", "--exclude=script/configs/exclude_xcode_deprecation.yaml", "--xcode-warnings-exceptions=script/configs/xcode_warnings_exceptions.yaml"]
   - name: native test
     script: .ci/scripts/tool_runner.sh
     args: ["native-test", "--macos"]
```
TODO: This line needs to be updated too

loic-sharma replied:
Updated

loic-sharma commented:
This is ready for another review! I removed the `--xcode-warnings-exceptions` flag from the `xcode-analyze` and switched the CI configs to use the `--exclude` flag instead.

stuartmorgan reviewed (approved):
LGTM
